kernel/safe_typing.mli

```diff
@@ -195,6 +195,19 @@ val export :
 val import : compiled_library -> Univ.ContextSet.t -> vodigest ->
   ModPath.t safe_transformer
 
+(** {6 Root extraction } *)
+
+type field =
+  | FModule of ModPath.t
+  | FModuleType of ModPath.t
+  | FConst of Constant.t
+  | FMind of MutInd.t
+
+val roots : safe_environment -> field list
```

Comment:
Two comments:
- with this approach I don't see the need of making `globals` transparent
- this API should take a filtering predicate, at least for the "module part". For example with `Search ... in Foo Bar` I limit my search to two modules, etc... That is, some filtering can be very cheap to do in this function (module name, label match) , while the more advanced one like "pattern matching" can be left to the caller. See https://coq.inria.fr/distrib/current/refman/proof-engine/ssreflect-proof-language.html#ssr-searching-tool

Reply:
What you propose shouldn't be in the kernel, much less in safe typing; also the `roots` API is pretty suboptimal when compared the maps, which at least have log access time.

I guess a way to evaluate the API is to answer the question "what kind of efficient indexers can you build _outside_ of the kernel". With `globals`, I can build some for a class of prefix-range based-search queries, with `roots`, none.

Reply:
Here's a change to providing a fold function, so you can do your filtering on the fly.
I don't know about indexing, I'm fine resetting to the first commit though.

Reply:
> Here's a change to providing a fold function, so you can do your filtering on the fly.
> I don't know about indexing, I'm fine resetting to the first commit though.

Fold is still linear so not too good for me; note that current maps in Coq allow to filter by range and work better for small result sets.

IMVHO from the point of view of `Search` and completion the new API is basically as useless as the old one, so I dunno if that's worth the addition to safe_env.

Reply:
>Fold is still linear so not too good for me; note that current maps in Coq allow to filter by range and work better for small result sets.

The hmaps that we have in the environment have a range filter? How?

>IMVHO from the point of view of Search and completion the new API is basically as useless as the old one, so I dunno if that's worth the addition to safe_env.

True

Reply:
> The hmaps that we have in the environment have a range filter? How?

Oh indeed, I happened to look at the `MPmap` which is indeed a `CMap`.